### PR TITLE
[WIP] Sort release cards

### PR DESCRIPTION
**Pull request by dobbymoodge**
Shuffle card positions to make sure they are ordered from smallest to greatest release number

### CODE REVIEW

File: lib/trello_helper.rb [hunk 1]
```diff
@@ -324,6 +324,18 @@ def board_lists(board, list_limit=max_lists_per_board)
     return lists
   end
 
+  def release_card_lists
+    release_states = NEXT_STATES.keys + BACKLOG_STATES.keys
+    boards.values.each do |board|
+      board_lists(board).each do |list|
+        labels = board_labels(board)
+        if release_states.include? list.name
+          yield(labels, list)
```
Comment (danmcp) [inline lib/trello_helper.rb]: I don't see a good reason to do this with a yield.

File: trello [hunk 2]
```diff
@@ -95,6 +95,47 @@ end
 
 trello = load_conf(TrelloHelper, CONFIG.trello, true)
 
+# Sort only the release cards by their release label release number
+# Sort only the cards in TrelloHelper::NEXT_STATES and TrelloHelper::BACKLOG_STATES
+command :organize_release_cards do |c|
+  c.syntax = "#{name}"
+  c.description = "rearrange board list cards sorted by release"
```
Comment (danmcp) [inline trello]: The other descriptions start with upper case.

File: trello [hunk 3]
```diff
@@ -95,6 +95,47 @@ end
...
+  c.action do |args, options|
+    trello.release_card_lists do |labels, list|
+      sorted_labels = labels.select do |label|
```
Comment (danmcp) [inline trello]: Could use some comments

File: trello [hunk 4]
```diff
@@ -95,6 +95,47 @@ end
 
 trello = load_conf(TrelloHelper, CONFIG.trello, true)
 
+# Sort only the release cards by their release label release number
+# Sort only the cards in TrelloHelper::NEXT_STATES and TrelloHelper::BACKLOG_STATES
+command :organize_release_cards do |c|
+  c.syntax = "#{name}"
```
Comment (danmcp) [inline trello]: dry run option?

File: trello [hunk 5]
```diff
@@ -95,6 +95,47 @@ end
...
+  c.description = "rearrange board list cards sorted by release"
+  c.action do |args, options|
+    trello.release_card_lists do |labels, list|
+      sorted_labels = labels.select do |label|
+        label.name =~ TrelloHelper::RELEASE_LABEL_REGEX
+      end.sort_by { |label| Gem::Version.new(label.name[TrelloHelper::RELEASE_LABEL_REGEX, 4]) }
+      label_id_to_release_number = Hash[sorted_labels.map{ |label| [label.id, Gem::Version.new(label.name[TrelloHelper::RELEASE_LABEL_REGEX, 4])] }]
```
Comment (danmcp) [inline trello]: I am missing where you are handling product.  Ex: targeted-online-3.4

File: trello [hunk 6]
```diff
@@ -95,6 +95,47 @@ end
...
+      release_cards = list_cards.select do |card|
+        card_labels = trello.card_labels(card)
+        card_labels.any? { |label| label_id_to_release_number.include? label.id if !label.nil? }
+      end
+      cards_by_pos = release_cards.sort_by { |card| card.pos }.map do |card|
+        card_release_label = trello.card_labels(card).find{ |label| label_id_to_release_number.include? label.id if !label.nil? }
+        [card, Gem::Version.new(card_release_label.name[TrelloHelper::RELEASE_LABEL_REGEX, 4]), card.pos]
+      end
+      cards_by_release = []
+      cards_by_pos.each { |card| cards_by_release << card.dup }
```
Comment (danmcp) [inline trello]: why dup?

File: trello [hunk 7]
```diff
@@ -95,6 +95,47 @@ end
...
+      card_updates = {}
+      cards_by_pos.sort_by! { |card| card[2] }
+      cards_by_release.sort_by! { |card| card[1] }
+      cards_by_pos.each_with_index do |card, idx|
+        if cards_by_release[idx][2] != card[2]
```
Comment (danmcp) [inline trello]: use index to be consistent

File: trello [hunk 8]
```diff
@@ -95,6 +95,47 @@ end
...
+        card_labels.any? { |label| label_id_to_release_number.include? label.id if !label.nil? }
+      end
+      cards_by_pos = release_cards.sort_by { |card| card.pos }.map do |card|
+        card_release_label = trello.card_labels(card).find{ |label| label_id_to_release_number.include? label.id if !label.nil? }
+        [card, Gem::Version.new(card_release_label.name[TrelloHelper::RELEASE_LABEL_REGEX, 4]), card.pos]
+      end
+      cards_by_release = []
+      cards_by_pos.each { |card| cards_by_release << card.dup }
+      card_updates = {}
+      cards_by_pos.sort_by! { |card| card[2] }
```
Comment (danmcp) [inline trello]: create cars for all the [2] stuff to make it more readable

File: trello (hunk 4 above)
Comment (dobbymoodge) [inline trello]: outdated diff, check again

File: trello [hunk 9]
```diff
@@ -95,6 +95,80 @@ end
 
 trello = load_conf(TrelloHelper, CONFIG.trello, true)
 
+# Sort only the release cards by their release label release number
+# Sort only the cards in TrelloHelper::NEXT_STATES and TrelloHelper::BACKLOG_STATES
+command :organize_release_cards do |c|
+
+  c.option "--dry-run", "Show what changes would be made without actually making them"
```
Comment (dobbymoodge) [inline trello]: @danmcp here

File: trello (hunk 8 above)
Comment (dobbymoodge) [inline trello]: using a map instead

File: lib/trello_helper.rb (hunk 1 above)
Comment (dobbymoodge) [inline lib/trello_helper.rb]: It saves a couple of indent levels later

File: trello (hunk 7 above)
Comment (dobbymoodge) [inline trello]: Done

File: lib/trello_helper.rb (hunk 1 above)
Comment (danmcp) [inline lib/trello_helper.rb]: It still doesn't feel like a good use case for a yield.

File: trello [hunk 10]
```diff
@@ -257,11 +338,11 @@ command :generate_release_json do |c|
 
   c.description = "Generate json for the cards in a release"
   c.action do |args, options|
-    release_cards = trello.release_cards(options.product, options.release)
+    release_label_to_cards = trello.release_label_to_cards(options.product, options.release)
```
Comment (danmcp) [inline trello]: What's this about?

Comment (dobbymoodge) [inline trello]: That's overzealous find-and-replace for you. :(

**Comment by dobbymoodge** [comment]
@stevekuznetsov PTAL. I'm 99% sure this code does the right thing, but it smells turrble. I could use your perspective here.

**Comment by dobbymoodge** [comment]
see https://github.com/openshift/sprint_tools/pull/60

**Pull request closed without merging**